README.md
## Overview

ConfigInit is a special container used to load (generate/restore) and dump (backup) the configuration and secrets.

## Versions

- Stable: `gluufederation/config-init:4.2.0_01`
- Unstable: `gluufederation/config-init:4.2.1_dev`

Refer to the [Changelog](./CHANGES.md) for details on new features, bug fixes, or older releases.

## Environment Variables

The following environment variables are supported by the container:

- `GLUU_CONFIG_ADAPTER`: The config backend adapter, can be `consul` (default) or `kubernetes`.
- `GLUU_CONFIG_CONSUL_HOST`: hostname or IP of Consul (default to `localhost`).
- `GLUU_CONFIG_CONSUL_PORT`: port of Consul (default to `8500`).
- `GLUU_CONFIG_CONSUL_CONSISTENCY`: Consul consistency mode (choose one of `default`, `consistent`, or `stale`). Default to `stale` mode.
- `GLUU_CONFIG_CONSUL_SCHEME`: supported Consul scheme (`http` or `https`).
- `GLUU_CONFIG_CONSUL_VERIFY`: whether to verify cert or not (default to `false`).
- `GLUU_CONFIG_CONSUL_CACERT_FILE`: path to Consul CA cert file (default to `/etc/certs/consul_ca.crt`). This file will be used if it exists and `GLUU_CONFIG_CONSUL_VERIFY` set to `true`.
- `GLUU_CONFIG_CONSUL_CERT_FILE`: path to Consul cert file (default to `/etc/certs/consul_client.crt`).
- `GLUU_CONFIG_CONSUL_KEY_FILE`: path to Consul key file (default to `/etc/certs/consul_client.key`).
- `GLUU_CONFIG_CONSUL_TOKEN_FILE`: path to file contains ACL token (default to `/etc/certs/consul_token`).
- `GLUU_CONFIG_KUBERNETES_NAMESPACE`: Kubernetes namespace (default to `default`).
- `GLUU_CONFIG_KUBERNETES_CONFIGMAP`: Kubernetes configmaps name (default to `gluu`).
- `GLUU_CONFIG_KUBERNETES_USE_KUBE_CONFIG`: Load credentials from `$HOME/.kube/config`, only useful for non-container environment (default to `false`).
- `GLUU_SECRET_ADAPTER`: The secrets adapter, can be `vault` or `kubernetes`.
- `GLUU_SECRET_VAULT_SCHEME`: supported Vault scheme (`http` or `https`).
- `GLUU_SECRET_VAULT_HOST`: hostname or IP of Vault (default to `localhost`).
- `GLUU_SECRET_VAULT_PORT`: port of Vault (default to `8200`).
- `GLUU_SECRET_VAULT_VERIFY`: whether to verify cert or not (default to `false`).
- `GLUU_SECRET_VAULT_ROLE_ID_FILE`: path to file contains Vault AppRole role ID (default to `/etc/certs/vault_role_id`).
- `GLUU_SECRET_VAULT_SECRET_ID_FILE`: path to file contains Vault AppRole secret ID (default to `/etc/certs/vault_secret_id`).
- `GLUU_SECRET_VAULT_CERT_FILE`: path to Vault cert file (default to `/etc/certs/vault_client.crt`).
- `GLUU_SECRET_VAULT_KEY_FILE`: path to Vault key file (default to `/etc/certs/vault_client.key`).
- `GLUU_SECRET_VAULT_CACERT_FILE`: path to Vault CA cert file (default to `/etc/certs/vault_ca.crt`). This file will be used if it exists and `GLUU_SECRET_VAULT_VERIFY` set to `true`.
- `GLUU_SECRET_KUBERNETES_NAMESPACE`: Kubernetes namespace (default to `default`).
- `GLUU_SECRET_KUBERNETES_CONFIGMAP`: Kubernetes secrets name (default to `gluu`).
- `GLUU_SECRET_KUBERNETES_USE_KUBE_CONFIG`: Load credentials from `$HOME/.kube/config`, only useful for non-container environment (default to `false`).
- `GLUU_WAIT_MAX_TIME`: How long the startup "health checks" should run (default to `300` seconds).
- `GLUU_WAIT_SLEEP_DURATION`: Delay between startup "health checks" (default to `10` seconds).
- `GLUU_OVERWRITE_ALL`: Overwrite all config (default to `false`).

## Commands

The following commands are supported by the container:

- `load`
- `dump`

### load

The load command can be used either to generate or restore config and secret for the cluster.

#### Docker

1. To generate the initial configuration and secret, create `/path/to/host/volume/generate.json` similar to example below:

    ```json
    {
        "hostname": "demoexample.gluu.org",
        "country_code": "US",
        "state": "TX",
        "city": "Austin",
        "admin_pw": "S3cr3t+pass",
        "ldap_pw": "S3cr3t+pass",
        "email": "s@gluu.local",
        "org_name": "Gluu Inc."
    }
    ```

1. Mount the volume into container:

    ```sh
    docker run \
        --rm \
        --network container:consul \
        -e GLUU_CONFIG_ADAPTER=consul \
        -e GLUU_CONFIG_CONSUL_HOST=consul \
        -e GLUU_SECRET_ADAPTER=vault \
        -e GLUU_SECRET_VAULT_HOST=vault \
        -v /path/to/host/volume:/opt/config-init/db \
        -v /path/to/vault_role_id.txt:/etc/certs/vault_role_id \
        -v /path/to/vault_secret_id.txt:/etc/certs/vault_secret_id \
        gluufederation/config-init:4.2.0_01 load
    ```

#### Kubernetes

1. To generate the initial configuration and secret, create `/path/to/host/volume/generate.json` similar to example below:

    ```json
    {
        "hostname": "demoexample.gluu.org",
        "country_code": "US",
        "state": "TX",
        "city": "Austin",
        "admin_pw": "S3cr3t+pass",
        "ldap_pw": "S3cr3t+pass",
        "email": "s@gluu.local",
        "org_name": "Gluu Inc."
    }
    ```

1. Create config map `config-generate-params`

   ```sh
   kubectl create cm config-generate-params --from-file=generate.json
   ```


1. Mount the configmap into container and apply the yaml:

    ```yaml
	apiVersion: batch/v1
	kind: Job
	metadata:
	  name: config-init-load-job
	spec:
	  template:
	    spec:
	      restartPolicy: Never
	      volumes:
	        - name: config-generate-params
	          configMap:
	            name: config-generate-params
	      containers:
	        - name: config-init-load
	          image: gluufederation/config-init:4.2.0_01
	          volumeMounts:
	            - mountPath: /opt/config-init/db/generate.json
	              name: config-generate-params
	              subPath: generate.json
	          envFrom:
	          - configMapRef:
	              name: config-cm
	          args: ["load"]
    ```

-   To restore configuration and secrets from a backup of `/path/to/host/volume/config.json` and `/path/to/host/volume/secret.json`: mount the directory as `/opt/config-init/db` inside the container:

1. Create config map `config-params` and `secret-params`:

   ```sh
   kubectl create cm config-params --from-file=config.json
   kubectl create cm secret-params --from-file=secret.json
   ```

1. Mount the configmap into container and apply the yaml:

    ```yaml
	apiVersion: batch/v1
	kind: Job
	metadata:
	  name: config-init-load-job
	spec:
	  template:
	    spec:
	      restartPolicy: Never
	      volumes:
	        - name: config-params
	          configMap:
	            name: config-params
	       	- name: secret-params
	          configMap:
	            name: secret-params
	      containers:
	        - name: config-init-load
	          image: gluufederation/config-init:4.2.0_01
	          volumeMounts:
	            - mountPath: /opt/config-init/db/config.json
	              name: config-params
	              subPath: config.json
	            - mountPath: /opt/config-init/db/secret.json
	              name: secret-params
	              subPath: secret.json
	          envFrom:
	          - configMapRef:
	              name: config-cm
	          args: ["load"]
	   ```


### dump

The dump command will dump all configuration and secrets from the backends saved into the `/opt/config-init/db/config.json` and `/opt/config-init/db/secret.json` files.

#### Docker

Please note that to dump this file into the host, mount a volume to the `/opt/config-init/db` directory as seen in the following example:

```sh
docker run \
    --rm \
    --network container:consul \
    -e GLUU_CONFIG_ADAPTER=consul \
    -e GLUU_CONFIG_CONSUL_HOST=consul \
    -e GLUU_SECRET_ADAPTER=vault \
    -e GLUU_SECRET_VAULT_HOST=vault \
    -v /path/to/host/volume:/opt/config-init/db \
    -v /path/to/vault_role_id.txt:/etc/certs/vault_role_id \
    -v /path/to/vault_secret_id.txt:/etc/certs/vault_secret_id \
    gluufederation/config-init:4.2.0_01 dump
```

#### Kubernetes

```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: config-init-load-job
spec:
  template:
    spec:
      restartPolicy: Never
      containers:
        - name: config-init-load
          image: gluufederation/config-init:4.2.0_01
          command:
            - /bin/sh
            - -c
            - |
                /app/scripts/entrypoint.sh dump
                sleep 300
          envFrom:
          - configMapRef:
              name: config-cm
```

Copy over the files to host

`kubectl cp config-init-load-job:opt/config-init/db .`


CHANGES.md
# Changelog

Here you can see an overview of changes between each release.

## Version 4.2.0_01

Released on July 18th, 2020.

* Added support for Gluu Server v4.2.
* Upraded to Java 11.
* Upgraded to Python3.

## Version 4.1.1_03

Released on June 5th, 2020.

* Upgraded `oxauth-client`.

## Version 4.1.1_02

Released on April 2nd, 2020.

* Upgraded `oxauth-client`.
* Added `curl` executable.

## Version 4.1.1_01

Released on March 24th, 2020.

* Conformed to Gluu Server v4.1.1.

## Version 4.1.0_01

Released on March 5th, 2020.

* Conformed to Gluu Server v4.1.
* Removed obsolete config and secret that previously used to automate Couchbase container.

## Version 4.0.1_10

Released on March 24th, 2020.

* Pre-populated SCIM RS cert alias. Ref: https://github.com/GluuFederation/docker-config-init/issues/16.
* Pre-populated API RS cert alias. Ref: https://github.com/GluuFederation/docker-config-init/issues/17.

## Version 4.0.1_09

Released on March 2nd, 2020.

* Pre-populated Passport RS cert alias.

## Version 4.0.1_08

Released on January 29th, 2020.

* Fixed hardcoded keystore password for oxTrust API.

## Version 4.0.1_07

Released on January 22nd, 2020.

* Upgraded `oxauth-client` to v4.0.Final.patch1 build at 2020-01-20.
* Added more supported keys algorithms.

## Version 4.0.1_06

Released on December 27th, 2019.

* Added credentials for SCIM test client.

## Version 4.0.1_05

Released on December 1st, 2019.

* Upgraded `oxauth-client` v4.0.1.Final build at 2019-11-30.

## Version 4.0.1_04

Released on November 19th, 2019.

* Upgraded `oxauth-client` v4.0.1.Final build at 2019-11-18.

## Version 4.0.1_03

Released on November 15th, 2019.

* Upgraded `oxauth-client` package.

## Version 4.0.1_02

Released on November 14th, 2019.

* Upgraded `pygluu-containerlib` to show connection issue with Couchbase explicitly.

## Version 4.0.1_01

Released on November 1st, 2019.

* Upgraded to Gluu Server 4.0.1.

## Version 4.0.0_01

Released on October 22nd, 2019.

* Upgraded to Gluu Server 4.0.
* Removed templates and static files to generate context for initial data.

## Version 3.1.6_02

Released on May 10th, 2019.

* Alpine upgraded to v3.9. Ref: https://github.com/GluuFederation/gluu-docker/issues/71.

## Version 3.1.6_01

Released on April 29th, 2019.

* Upgraded to Gluu Server 3.1.6.

## Version 3.1.5_05

Released on May 10th, 2019.

* Alpine upgraded to v3.9. Ref: https://github.com/GluuFederation/gluu-docker/issues/71.

## Version 3.1.5_04

Released on April 23rd, 2019.

* Changed default expiration time for oxAuth keys.

## Version 3.1.5_03

Released on April 9th, 2019.

* Added license info on container startup.
* Added new command `migrate` to migrate some keys from config to secret backend.

## Version 3.1.5_02

Released on March 28th, 2019.

* Added `wait_for` in `load` and `dump` commands.

## Version 3.1.5_01

Released on March 23rd, 2019.

* Upgraded to Gluu Server 3.1.5.

## Version 3.1.4_04

Released on April 4th, 2019.

* Added license info during container run.

## Version 3.1.4_03

Released on December 2nd, 2018.

* Removed Casa scripts from distribution.

## Version 3.1.4_02

Released on November 28th, 2018.

* Fixed issue with single-quotation when generating certificate.

## Version 3.1.4_01

Released on November 12th, 2018.

* Upgraded to Gluu Server 3.1.4.

## Version 3.1.3_06

Released on September 18th, 2018.

* Changed base image to use Alpine 3.8.1.

## Version 3.1.3_05

Released on September 12th, 2018.

* Added feature to connect to secure Consul (HTTPS).

## Version 3.1.3_04

Released on September 6th, 2018.

* Added feature to guard existing keys being overwritten when executing `generate` and `load` commands.
* Added environment variable to force rewrite all keys when executing `generate` and `load` commands.

## Version 3.1.3_03

Released on August 3rd, 2018.

* Added optional params for baseInum, inumOrg, and inumAppliance.

## Version 3.1.3_02

Released on July 19th, 2018.

* Added wrapper to manage config via Consul KV or Kubernetes configmap.
* Added feature to dump config into JSON file when running `generate` command.

## Version 3.1.3_01

Released on June 6th, 2018.

* Upgraded to Gluu Server 3.1.3.

## Version 3.1.2_01

Released on June 6th, 2018.

* Upgraded to Gluu Server 3.1.2.

## Version 3.1.1_rev1.0.0-beta3

Released on October 25th, 2017.

* Fixed work_phone dynamic script.

## Version 3.1.1_rev1.0.0-beta2

Released on October 11th, 2017.

* Added default person OC.

## Version 3.1.1_rev1.0.0-beta1

Released on October 6th, 2017.

* Migrated to Gluu Server 3.1.1.

## Version 3.0.1_rev1.0.0-beta6

Released on September 29th, 2017.

* Added feature to use existing `oxauth-keys.jks` file (for migrating external LDAP/CE version).

## Version 3.0.1_rev1.0.0-beta5

Released on September 19th, 2017.

* Added options to input external encoded salt, ox LDAP password, and inum appliance (for migrating external LDAP/CE version).

## Version 3.0.1_rev1.0.0-beta4

Released on August 23rd, 2017.

* Added patches to prevent accidentally updating keys when entrypoint is re-executed.

## Version 3.0.1_rev1.0.0-beta3

Released on July 20th, 2017.

* Added config to store first instance of LDAP


scripts/gluulib.py
import base64
import json
import logging
import os
from collections import namedtuple

import hvac
import six
import kubernetes.client
import kubernetes.config
from consul import Consul

logger = logging.getLogger("gluulib")
logger.setLevel(logging.INFO)
ch = logging.StreamHandler()
fmt = logging.Formatter('%(levelname)s - %(asctime)s - %(message)s')
ch.setFormatter(fmt)
logger.addHandler(ch)


def as_boolean(val, default=False):
    """Converts value as boolean.
    """
    truthy = set(('t', 'T', 'true', 'True', 'TRUE', '1', 1, True))
    falsy = set(('f', 'F', 'false', 'False', 'FALSE', '0', 0, False))

    if val in truthy:
        return True
    if val in falsy:
        return False
    return default


def safe_value(value):
    if not isinstance(value, (six.string_types, six.binary_type)):
        value = json.dumps(value)
    return value


class BaseConfig(object):
    """Base class for config adapter. Must be sub-classed per
    implementation details.
    """
    type = "config"

    def get(self, key, default=None):
        """Get specific config.

        Subclass __MUST__ implement this method.
        """
        raise NotImplementedError

    def set(self, key, value):
        """Set specific config.

        Subclass __MUST__ implement this method.
        """
        raise NotImplementedError

    def delete(self, key):
        """Deletes specific config.

        Subclass __MUST__ implement this method.
        """
        raise NotImplementedError

    def all(self):
        """Get all config as ``dict`` type.

        Subclass __MUST__ implement this method.
        """
        raise NotImplementedError


class ConsulConfig(BaseConfig):
    def __init__(self):
        # collects all env vars prefixed with `GLUU_CONFIG_CONSUL_`,
        # for example `GLUU_CONFIG_CONSUL_HOST=localhost`
        self.settings = {
            k: v for k, v in os.environ.iteritems()
            if k.isupper() and k.startswith("GLUU_CONFIG_CONSUL_")
        }

        self.settings.setdefault(
            "GLUU_CONFIG_CONSUL_HOST",
            # backward-compat with Gluu Server v3.1.4
            os.environ.get("GLUU_CONSUL_HOST", "localhost"),
        )

        self.settings.setdefault(
            "GLUU_CONFIG_CONSUL_PORT",
            # backward-compat with Gluu Server v3.1.4
            os.environ.get("GLUU_CONSUL_PORT", 8500),
        )

        self.settings.setdefault(
            "GLUU_CONFIG_CONSUL_CONSISTENCY",
            # backward-compat with Gluu Server v3.1.4
            os.environ.get("GLUU_CONSUL_CONSISTENCY", "stale"),
        )

        self.settings.setdefault(
            "GLUU_CONFIG_CONSUL_SCHEME",
            # backward-compat with Gluu Server v3.1.4
            os.environ.get("GLUU_CONSUL_SCHEME", "http"),
        )

        self.settings.setdefault(
            "GLUU_CONFIG_CONSUL_VERIFY",
            # backward-compat with Gluu Server v3.1.4
            os.environ.get("GLUU_CONSUL_VERIFY", False),
        )

        self.settings.setdefault(
            "GLUU_CONFIG_CONSUL_CACERT_FILE",
            # backward-compat with Gluu Server v3.1.4
            os.environ.get("GLUU_CONSUL_CACERT_FILE",
                           "/etc/certs/consul_ca.crt"),
        )

        self.settings.setdefault(
            "GLUU_CONFIG_CONSUL_CERT_FILE",
            # backward-compat with Gluu Server v3.1.4
            os.environ.get("GLUU_CONSUL_CERT_FILE",
                           "/etc/certs/consul_client.crt"),
        )

        self.settings.setdefault(
            "GLUU_CONFIG_CONSUL_KEY_FILE",
            # backward-compat with Gluu Server v3.1.4
            os.environ.get("GLUU_CONSUL_KEY_FILE",
                           "/etc/certs/consul_client.key"),
        )

        self.settings.setdefault(
            "GLUU_CONFIG_CONSUL_TOKEN_FILE",
            # backward-compat with Gluu Server v3.1.4
            os.environ.get("GLUU_CONSUL_TOKEN_FILE", "/etc/certs/consul_token"),
        )

        self.prefix = "gluu/config/"
        token = None
        cert = None
        verify = False

        if os.path.isfile(self.settings["GLUU_CONFIG_CONSUL_TOKEN_FILE"]):
            with open(self.settings["GLUU_CONFIG_CONSUL_TOKEN_FILE"]) as fr:
                token = fr.read().strip()

        if self.settings["GLUU_CONFIG_CONSUL_SCHEME"] == "https":
            verify = as_boolean(self.settings["GLUU_CONFIG_CONSUL_VERIFY"])

            # verify using CA cert (if any)
            if all([verify,
                    os.path.isfile(self.settings["GLUU_CONFIG_CONSUL_CACERT_FILE"])]):
                verify = self.settings["GLUU_CONFIG_CONSUL_CACERT_FILE"]

            if all([os.path.isfile(self.settings["GLUU_CONFIG_CONSUL_CERT_FILE"]),
                    os.path.isfile(self.settings["GLUU_CONFIG_CONSUL_KEY_FILE"])]):
                cert = (self.settings["GLUU_CONFIG_CONSUL_CERT_FILE"],
                        self.settings["GLUU_CONFIG_CONSUL_KEY_FILE"])

        self._request_warning(self.settings["GLUU_CONFIG_CONSUL_SCHEME"], verify)

        self.client = Consul(
            host=self.settings["GLUU_CONFIG_CONSUL_HOST"],
            port=self.settings["GLUU_CONFIG_CONSUL_PORT"],
            token=token,
            scheme=self.settings["GLUU_CONFIG_CONSUL_SCHEME"],
            consistency=self.settings["GLUU_CONFIG_CONSUL_CONSISTENCY"],
            verify=verify,
            cert=cert,
        )

    def _merge_path(self, key):
        """Add prefix to the key.
        """
        return "".join([self.prefix, key])

    def _unmerge_path(self, key):
        """Remove prefix from the key.
        """
        return key[len(self.prefix):]

    def get(self, key, default=None):
        _, result = self.client.kv.get(self._merge_path(key))
        if not result:
            return default
        return result["Value"]

    def set(self, key, value):
        return self.client.kv.put(self._merge_path(key),
                                  safe_value(value))

    def find(self, key):
        _, resultset = self.client.kv.get(self._merge_path(key),
                                          recurse=True)

        if not resultset:
            return {}

        return {
            self._unmerge_path(item["Key"]): item["Value"]
            for item in resultset
        }

    def all(self):
        return self.find("")

    def delete(self, key):
        return self.client.kv.delete(self._merge_path(key))

    def _request_warning(self, scheme, verify):
        if scheme == "https" and verify is False:
            import urllib3
            urllib3.disable_warnings()
            logger.warn(
                "All requests to Consul will be unverified. "
                "Please adjust GLUU_CONFIG_CONSUL_SCHEME and "
                "GLUU_CONFIG_CONSUL_VERIFY environment variables."
            )


class KubernetesConfig(BaseConfig):
    def __init__(self):
        self.settings = {
            k: v for k, v in os.environ.iteritems()
            if k.isupper() and k.startswith("GLUU_CONFIG_KUBERNETES_")
        }
        self.settings.setdefault(
            "GLUU_CONFIG_KUBERNETES_NAMESPACE",
            # backward-compat with Gluu Server v3.1.4
            os.environ.get("GLUU_KUBERNETES_NAMESPACE", "default"),
        )

        self.settings.setdefault(
            "GLUU_CONFIG_KUBERNETES_CONFIGMAP",
            # backward-compat with Gluu Server v3.1.4
            os.environ.get("GLUU_KUBERNETES_CONFIGMAP", "gluu"),
        )

        self.settings.setdefault(
            "GLUU_CONFIG_KUBERNETES_USE_KUBE_CONFIG",
            False
        )

        if as_boolean(self.settings["GLUU_CONFIG_KUBERNETES_USE_KUBE_CONFIG"]):
            kubernetes.config.load_kube_config()
        else:
            kubernetes.config.load_incluster_config()

        self.client = kubernetes.client.CoreV1Api()
        self.name_exists = False

    def get(self, key, default=None):
        result = self.all()
        return result.get(key, default)

    def _prepare_configmap(self):
        # create a configmap name if not exist
        if not self.name_exists:
            try:
                self.client.read_namespaced_config_map(
                    self.settings["GLUU_CONFIG_KUBERNETES_CONFIGMAP"],
                    self.settings["GLUU_CONFIG_KUBERNETES_NAMESPACE"])
                self.name_exists = True
            except kubernetes.client.rest.ApiException as exc:
                if exc.status == 404:
                    # create the configmaps name
                    body = {
                        "kind": "ConfigMap",
                        "apiVersion": "v1",
                        "metadata": {
                            "name": self.settings["GLUU_CONFIG_KUBERNETES_CONFIGMAP"],
                        },
                        "data": {},
                    }
                    created = self.client.create_namespaced_config_map(
                        self.settings["GLUU_CONFIG_KUBERNETES_NAMESPACE"],
                        body)
                    if created:
                        self.name_exists = True
                else:
                    raise

    def set(self, key, value):
        self._prepare_configmap()
        body = {
            "kind": "ConfigMap",
            "apiVersion": "v1",
            "metadata": {
                "name": self.settings["GLUU_CONFIG_KUBERNETES_CONFIGMAP"],
            },
            "data": {
                key: safe_value(value),
            }
        }
        return self.client.patch_namespaced_config_map(
            self.settings["GLUU_CONFIG_KUBERNETES_CONFIGMAP"],
            self.settings["GLUU_CONFIG_KUBERNETES_NAMESPACE"],
            body=body)

    def all(self):
        self._prepare_configmap()
        result = self.client.read_namespaced_config_map(
            self.settings["GLUU_CONFIG_KUBERNETES_CONFIGMAP"],
            self.settings["GLUU_CONFIG_KUBERNETES_NAMESPACE"])
        return result.data or {}

    def delete(self, key):
        result = self.all()
        result.pop(key, None)

        body = {
            "kind": "ConfigMap",
            "apiVersion": "v1",
            "metadata": {
                "name": self.settings["GLUU_CONFIG_KUBERNETES_CONFIGMAP"],
            },
            "data": result,
        }
        return self.client.replace_namespaced_config_map(
            self.settings["GLUU_CONFIG_KUBERNETES_CONFIGMAP"],
            self.settings["GLUU_CONFIG_KUBERNETES_NAMESPACE"],
            body=body)


class ConfigManager(object):
    def __init__(self):
        _adapter = os.environ.get(
            "GLUU_CONFIG_ADAPTER",
            "consul",
        )
        if _adapter == "consul":
            self.adapter = ConsulConfig()
        elif _adapter == "kubernetes":
            self.adapter = KubernetesConfig()
        else:
            self.adapter = None

    def get(self, key, default=None):
        return self.adapter.get(key, default)

    def set(self, key, value):
        return self.adapter.set(key, value)

    def all(self):
        return self.adapter.all()

    def delete(self, key):
        return self.adapter.delete(key)


class BaseSecret(object):
    """Base class for secret adapter. Must be sub-classed per
    implementation details.
    """
    type = "secret"

    def get(self, key, default=None):
        """Get specific secret.

        Subclass __MUST__ implement this method.
        """
        raise NotImplementedError

    def set(self, key, value):
        """Set specific secret.

        Subclass __MUST__ implement this method.
        """
        raise NotImplementedError

    def all(self):
        """Get all secrets as ``dict`` type.

        Subclass __MUST__ implement this method.
        """
        raise NotImplementedError


class VaultSecret(BaseSecret):
    def __init__(self):
        self.settings = {
            k: v for k, v in os.environ.iteritems()
            if k.isupper() and k.startswith("GLUU_SECRET_VAULT_")
        }
        self.settings.setdefault(
            "GLUU_SECRET_VAULT_HOST",
            "localhost",
        )
        self.settings.setdefault(
            "GLUU_SECRET_VAULT_PORT",
            8200,
        )
        self.settings.setdefault(
            "GLUU_SECRET_VAULT_SCHEME",
            "http",
        )
        self.settings.setdefault(
            "GLUU_SECRET_VAULT_VERIFY",
            False,
        )
        self.settings.setdefault(
            "GLUU_SECRET_VAULT_ROLE_ID_FILE",
            "/etc/certs/vault_role_id",
        ),
        self.settings.setdefault(
            "GLUU_SECRET_VAULT_SECRET_ID_FILE",
            "/etc/certs/vault_secret_id",
        )
        self.settings.setdefault(
            "GLUU_SECRET_VAULT_CERT_FILE",
            "/etc/certs/vault_client.crt",
        )
        self.settings.setdefault(
            "GLUU_SECRET_VAULT_KEY_FILE",
            "/etc/certs/vault_client.key",
        )
        self.settings.setdefault(
            "GLUU_SECRET_VAULT_CACERT_FILE",
            "/etc/certs/vault_ca.crt",
        )

        cert = None
        verify = False

        if self.settings["GLUU_SECRET_VAULT_SCHEME"] == "https":
            verify = as_boolean(self.settings["GLUU_SECRET_VAULT_VERIFY"])

            # verify using CA cert (if any)
            if all([verify,
                    os.path.isfile(self.settings["GLUU_SECRET_VAULT_CACERT_FILE"])]):
                verify = self.settings["GLUU_SECRET_VAULT_CACERT_FILE"]

            if all([os.path.isfile(self.settings["GLUU_SECRET_VAULT_CERT_FILE"]),
                    os.path.isfile(self.settings["GLUU_SECRET_VAULT_KEY_FILE"])]):
                cert = (self.settings["GLUU_SECRET_VAULT_CERT_FILE"],
                        self.settings["GLUU_SECRET_VAULT_KEY_FILE"])

        self._request_warning(self.settings["GLUU_SECRET_VAULT_SCHEME"], verify)

        self.client = hvac.Client(
            url="{}://{}:{}".format(
                self.settings["GLUU_SECRET_VAULT_SCHEME"],
                self.settings["GLUU_SECRET_VAULT_HOST"],
                self.settings["GLUU_SECRET_VAULT_PORT"],
            ),
            cert=cert,
            verify=verify,
        )
        self.prefix = "secret/gluu"

    @property
    def role_id(self):
        try:
            with open(self.settings["GLUU_SECRET_VAULT_ROLE_ID_FILE"]) as f:
                role_id = f.read()
        except IOError:
            role_id = ""
        return role_id

    @property
    def secret_id(self):
        try:
            with open(self.settings["GLUU_SECRET_VAULT_SECRET_ID_FILE"]) as f:
                secret_id = f.read()
        except IOError:
            secret_id = ""
        return secret_id

    def _authenticate(self):
        if self.client.is_authenticated():
            return

        creds = self.client.auth_approle(self.role_id, self.secret_id, use_token=False)
        self.client.token = creds["auth"]["client_token"]

    def get(self, key, default=None):
        self._authenticate()
        sc = self.client.read("{}/{}".format(self.prefix, key))
        if not sc:
            return default
        return sc["data"]["value"]

    def set(self, key, value):
        self._authenticate()
        val = {"value": value}

        # hvac.v1.Client.write checks for status code 200,
        # but Vault HTTP API returns 204 if request succeeded;
        # hence we're using lower level of `hvac.v1.Client` API to set key-val
        response = self.client._adapter.post('/v1/{0}/{1}'.format(self.prefix, key), json=val)
        return response.status_code == 204

    def all(self):
        self._authenticate()
        result = self.client.list(self.prefix)
        return {key: self.get(key) for key in result["data"]["keys"]}

    def _request_warning(self, scheme, verify):
        if scheme == "https" and verify is False:
            import urllib3
            urllib3.disable_warnings()
            logger.warn(
                "All requests to Vault will be unverified. "
                "Please adjust GLUU_SECRET_VAULT_SCHEME and "
                "GLUU_SECRET_VAULT_VERIFY environment variables."
            )


class KubernetesSecret(BaseSecret):
    def __init__(self):
        self.settings = {
            k: v for k, v in os.environ.iteritems()
            if k.isupper() and k.startswith("GLUU_SECRET_KUBERNETES_")
        }
        self.settings.setdefault(
            "GLUU_SECRET_KUBERNETES_NAMESPACE",
            "default",
        )
        self.settings.setdefault(
            "GLUU_SECRET_KUBERNETES_SECRET",
            "gluu",
        )
        self.settings.setdefault(
            "GLUU_SECRET_KUBERNETES_USE_KUBE_CONFIG",
            False
        )

        if as_boolean(self.settings["GLUU_SECRET_KUBERNETES_USE_KUBE_CONFIG"]):
            kubernetes.config.load_kube_config()
        else:
            kubernetes.config.load_incluster_config()
        self.client = kubernetes.client.CoreV1Api()
        self.name_exists = False

    def get(self, key, default=None):
        result = self.all()
        if key in result:
            return base64.b64decode(result[key])
        return default

    def _prepare_secret(self):
        # create a secret name if not exist
        if not self.name_exists:
            try:
                self.client.read_namespaced_secret(
                    self.settings["GLUU_SECRET_KUBERNETES_SECRET"],
                    self.settings["GLUU_SECRET_KUBERNETES_NAMESPACE"])
                self.name_exists = True
            except kubernetes.client.rest.ApiException as exc:
                if exc.status == 404:
                    # create the secrets name
                    body = {
                        "kind": "Secret",
                        "apiVersion": "v1",
                        "metadata": {
                            "name": self.settings["GLUU_SECRET_KUBERNETES_SECRET"],
                        },
                        "data": {},
                    }
                    created = self.client.create_namespaced_secret(
                        self.settings["GLUU_SECRET_KUBERNETES_NAMESPACE"],
                        body)
                    if created:
                        self.name_exists = True
                else:
                    raise

    def set(self, key, value):
        self._prepare_secret()
        body = {
            "kind": "Secret",
            "apiVersion": "v1",
            "metadata": {
                "name": self.settings["GLUU_SECRET_KUBERNETES_SECRET"],
            },
            "data": {
                key: base64.b64encode(value),
            }
        }
        return self.client.patch_namespaced_secret(
            self.settings["GLUU_SECRET_KUBERNETES_SECRET"],
            self.settings["GLUU_SECRET_KUBERNETES_NAMESPACE"],
            body=body)

    def all(self):
        self._prepare_secret()
        result = self.client.read_namespaced_secret(
            self.settings["GLUU_SECRET_KUBERNETES_SECRET"],
            self.settings["GLUU_SECRET_KUBERNETES_NAMESPACE"])
        return result.data or {}


class SecretManager(object):
    def __init__(self, config_adapter=None):
        _adapter = os.environ.get(
            "GLUU_SECRET_ADAPTER",
            "vault",
        )
        if _adapter == "vault":
            self.adapter = VaultSecret()
        elif _adapter == "kubernetes":
            self.adapter = KubernetesSecret()
        else:
            self.adapter = None

        # backward-compat
        self.config_adapter = config_adapter

    def _get_compat(self, key, default=None):
        sc = self.adapter.get(key, default)

        if not sc and self.config_adapter:
            # tries to get from old config
            sc = self.config_adapter.get(key, default)

        if not sc:
            # fallback to default
            sc = default
        return sc

    def get(self, key, default=None):
        return self._get_compat(key, default)

    def set(self, key, value):
        return self.adapter.set(key, value)

    def all(self):
        return self.adapter.all()


def get_manager():
    """Convenient function to get manager instances.
    """
    obj = namedtuple("Manager", "config secret")
    config_mgr = ConfigManager()
    secret_mgr = SecretManager(config_mgr)
    return obj(config=config_mgr, secret=secret_mgr)
